dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/security/UserAccountController.java

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@OpenApi.Tags({"user", "login", "account"})
+@RestController
+@RequestMapping("/auth")
+@RequiredArgsConstructor
+@ApiVersion({DhisApiVersion.DEFAULT, DhisApiVersion.ALL})
+@Slf4j
+public class UserAccountController {
+
+  private final UserService userService;
+  private final SystemSettingManager systemSettingManager;
+  private final PasswordValidationService passwordValidationService;
+
+  @PostMapping("/forgotPassword")
+  @ResponseStatus(HttpStatus.OK)
+  public void forgotPassword(
+      HttpServletRequest request, @RequestBody ForgotPasswordRequest forgotPasswordRequest)
+      throws NotFoundException, ConflictException, ForbiddenException {
+
+    String username = forgotPasswordRequest.getUsername();
+
+    if (userService.isRecoveryLocked(username)) {
+      throw new ForbiddenException(
+          "The account recovery operation for the given user is temporarily locked due to too "
+              + "many calls to this endpoint in the last '"
+              + RECOVERY_LOCKOUT_MINS
+              + "' minutes. Username:"
+              + username);
+    } else {
+      userService.registerRecoveryAttempt(username);
+    }
+
+    User user = userService.getUserByUsername(username);
+
+    if (user == null) {
+      throw new NotFoundException("User does not exist: " + username);
+    }
+
+    ErrorCode errorCode = userService.validateRestore(user);
+
+    if (errorCode != null) {
+      throw new IllegalQueryException(errorCode);
+    }
+
+    if (!userService.sendRestoreOrInviteMessage(
+        user, ContextUtils.getContextPath(request), RestoreOptions.RECOVER_PASSWORD_OPTION)) {
+      throw new ConflictException("Account could not be recovered");
+    }
+
+    log.info("Recovery message sent for user: " + username);
+  }
+
+  @PostMapping("/passwordReset")
+  @ResponseStatus(HttpStatus.OK)
+  public void resetPassword(@RequestBody ResetPasswordRequest resetRequest)
+      throws ConflictException, BadRequestException {
+
+    String token = resetRequest.getResetToken();
+    String newPassword = resetRequest.getNewPassword();
+
+    String[] idAndRestoreToken = userService.decodeEncodedTokens(token);
+    String idToken = idAndRestoreToken[0];
+
+    User user = userService.getUserByIdToken(idToken);
+    if (user == null || idAndRestoreToken.length < 2) {
+      throw new ConflictException("Account recovery failed");
+    }
+    String restoreToken = idAndRestoreToken[1];
+
+    if (!systemSettingManager.accountRecoveryEnabled()) {
+      throw new ConflictException("Account recovery is not enabled");
+    }
+
+    if (newPassword.trim().equals(user.getUsername())) {
+      throw new BadRequestException("Password cannot be equal to username");
+    }
+
+    CredentialsInfo credentialsInfo =
+        new CredentialsInfo(
+            user.getUsername(), newPassword, user.getEmail() != null ? user.getEmail() : "", false);
```

Comment:
Use commons `StringUtils` > `trimToEmpty` for email instead of ternary operator.